src/eve/pattern_matching.py

```diff
@@ -0,0 +1,139 @@
+
+        If `raise_` is specified raises an exception with all differences found.
+        """
+        diffs = list(get_differences(self, other))
+        if raise_ and len(diffs) != 0:
+            diffs_str = "\n  ".join([f"  {self.cls.__name__}{path}: {msg}" for path, msg in diffs])
+            raise ValueError(f"Object and pattern don't match:\n  {diffs_str}")
+        return len(diffs) == 0
+
+    def __str__(self) -> str:
+        attrs_str = ", ".join([f"{str(k)}={str(v)}" for k, v in self.attrs.items()])
+        return f"{self.cls.__name__}({attrs_str})"
+
+
+@dataclass(frozen=True)
+class ObjectPatternConstructor:
+    cls: type
+
+    def __call__(self, **kwargs: Any) -> ObjectPattern:
+        return ObjectPattern(self.cls, kwargs)
+
+
+def _get_differences_object_pattern(
+    a: ObjectPattern, b: Any, path: str = ""
+) -> Iterator[tuple[str, str]]:
+    if not isinstance(b, a.cls):
+        yield (
+            path,
+            f"Expected an instance of class {a.cls.__name__}, but got {type(b).__name__}",
+        )
+    else:
+        for k in a.attrs.keys():
+            if not hasattr(b, k):
+                yield (path, f"Value has no attribute {k}.")
+            else:
+                for diff in get_differences(a.attrs[k], getattr(b, k), path=f"{path}.{k}"):
+                    yield diff
+
+
+def _get_differences_list(a: list, b: Any, path: str = "") -> Iterator[tuple[str, str]]:
+    if not isinstance(b, list):
+        yield (path, f"Expected list, but got {type(b).__name__}")
+    elif len(a) != len(b):
+        yield (path, f"Expected list of length {len(a)}, but got length {len(b)}")
+    else:
+        for i, (el_a, el_b) in enumerate(zip(a, b)):
+            for diff in get_differences(el_a, el_b, path=f"{path}[{i}]"):
+                yield diff
+
+
+def _get_differences_dict(a: dict, b: Any, path: str = "") -> Iterator[tuple[str, str]]:
+    if not isinstance(b, dict):
+        yield (path, f"Expected dict, but got {type(b).__name__}")
+    elif set(a.keys()) != set(b.keys()):
+        a_min_b = set(a.keys()).difference(b.keys())
+        b_min_a = set(b.keys()).difference(a.keys())
+        if a_min_b:
+            missing_keys_str = "`" + "`, `".join(map(str, a_min_b)) + "`"
+            yield (
+                path,
+                f"Expected dictionary with keys `{'`, `'.join(map(str, a.keys()))}`, but the following keys are missing: {missing_keys_str}",
+            )
+        if b_min_a:
+            extra_keys_str = "`" + "`, `".join(map(str, b_min_a)) + "`"
+            yield (
+                path,
+                f"Expected dictionary with keys `{'`, `'.join(map(str, a.keys()))}`, but the following keys are extra: {extra_keys_str}",
+            )
+    else:
+        for k, v_a, v_b in zip(a.keys(), a.values(), b.values()):
+            yield from get_differences(v_a, v_b, path=f'{path}["{k}"]')
+
+
+def get_differences(a: Any, b: Any, path: str = "") -> Iterator[tuple[str, str]]:
```

Comment:
I would protect the `path` argument since it's an internal argument not meant to be used in the public interface of the function.

```suggestion
def get_differences(a: Any, b: Any, *, __path: str = "") -> Iterator[tuple[str, str]]:
```

Reply:
It is explicitly meant as a public argument. This is actually used in one of the tests:
```
@pytest.mark.parametrize("name,test_data", test_data)
def test_all(name, test_data):
    a, b, expected_diff = test_data["a"], test_data["b"], test_data["expected_diff"]
    diff = list(get_differences(a, b, path="a"))
```
The resulting messages contain valid python expressions to access the respective value and are very handy to be used inside the debugger.